concretecloud/chirp
# Change to portable network proto using a serializer

This commit is missing the following:

* Unittests

I request the following review:

* Please check if my pointer arithmetic style is portable, since
  paranoid C doesn't allow "void* a = x; a += 3;" we have to move
  through the buffer in a "char" context and then cast to void*

* Please do a network research to see if you come to the same conclusion
  as I.

* Please review the actual code (all: serializer.rg.h, sosterwalder: everything)

About the problem:

It seems that structs are mostly portable, but not in the least guaranteed
to be portable.
https://stackoverflow.com/questions/45116212/are-packed-structs-portable

What is currently done:

* Assume the layout and size of struct is the same on the target
  platforms

* Change every field that goes to the network to network-order and vice
  versa

What should be done:

* Change every field that goes to the network to network-order and vice
  versa

* Layout and read the data manually, meaning write a manual
  (de-)serializer, that is guaranteed to do the same layout on every
  platform

* Don't use sizeof the message, but the size we learned by designing the
  serializer (a constant)

No, using msgpack or anything similar is not a good idea, since most
implementation are too complex and often not portable (not the
serialization format is not portable, but the implementation).
